nodebuilder/header/data_commitment.go

```diff
@@ -0,0 +1,242 @@
+	}
+	heightsRange := end - start
+	if heightsRange > uint64(dataCommitmentBlocksLimit) {
+		return fmt.Errorf("the query exceeds the limit of allowed blocks %d", dataCommitmentBlocksLimit)
+	}
+
+	currentHeader, err := s.NetworkHead(ctx)
+	if err != nil {
+		return err
+	}
+	// the data commitment range is end exclusive
+	if end > currentHeader.Height()+1 {
+		return fmt.Errorf(
+			"end block %d is higher than current chain height %d",
+			end,
+			currentHeader.Height(),
+		)
+	}
+
+	currentLocalHeader, err := s.LocalHead(ctx)
+	if err != nil {
+		return err
+	}
+	// the data commitment range is end exclusive
+	if end > currentLocalHeader.Height()+1 {
+		return fmt.Errorf(
+			"end block %d is higher than local chain height %d. Wait for the node until it syncs up to %d",
+			end,
+			currentLocalHeader.Height(),
+			end,
+		)
+	}
+	return nil
+}
+
+// hashDataRootTuples hashes a list of blocks data root tuples, i.e., height, data root and square
+// size, then returns their merkle root.
+func hashDataRootTuples(tuples []DataRootTuple) ([]byte, error) {
+	if len(tuples) == 0 {
+		return nil, fmt.Errorf("cannot hash an empty list of data root tuples")
+	}
+	dataRootEncodedTuples := make([][]byte, 0, len(tuples))
+	for _, tuple := range tuples {
+		encodedTuple, err := EncodeDataRootTuple(
+			tuple.height,
+			tuple.dataRoot,
+		)
+		if err != nil {
+			return nil, err
+		}
+		dataRootEncodedTuples = append(dataRootEncodedTuples, encodedTuple)
+	}
+	root := merkle.HashFromByteSlices(dataRootEncodedTuples)
+	return root, nil
+}
+
+// validateDataRootInclusionProofRequest validates the request to generate a data root
+// inclusion proof.
+func (s *Service) validateDataRootInclusionProofRequest(
+	ctx context.Context,
+	height, start, end uint64,
+) error {
+	err := s.validateDataCommitmentRange(ctx, start, end)
+	if err != nil {
+		return err
+	}
+	if height < start || height >= end {
+		return fmt.Errorf(
+			"height %d should be in the end exclusive interval first_block %d last_block %d",
+			height,
+			start,
+			end,
+		)
+	}
+	return nil
+}
+
+// proveDataRootTuples returns the merkle inclusion proof for a height.
+func proveDataRootTuples(tuples []DataRootTuple, height int64) (*merkle.Proof, error) {
+	if len(tuples) == 0 {
+		return nil, fmt.Errorf("cannot prove an empty list of tuples")
+	}
+	if height < 0 {
+		return nil, fmt.Errorf("cannot prove a strictly negative height %d", height)
+	}
+	currentHeight := tuples[0].height - 1
+	for _, tuple := range tuples {
+		if tuple.height != currentHeight+1 {
+			return nil, fmt.Errorf("the provided tuples are not consecutive %d vs %d", currentHeight, tuple.height)
+		}
+		currentHeight++
+	}
+	dataRootEncodedTuples := make([][]byte, 0, len(tuples))
+	for _, tuple := range tuples {
+		encodedTuple, err := EncodeDataRootTuple(
+			tuple.height,
+			tuple.dataRoot,
+		)
+		if err != nil {
+			return nil, err
+		}
+		dataRootEncodedTuples = append(dataRootEncodedTuples, encodedTuple)
```

Comment:
why can't this also be done in `fetchDataRootTuples` ? both callers of `fetchDataRootTuples` want an encoded version of the tuples.

Reply:
you mean to reduce the duplicated code? I think it's fine since we use the information from the non encoded data root tuple to know the starting height of the range. Otherwise, we will need to pass it as a parameter.

Anyway, I am fine with both, let me know what you prefer

Reply:
I'd rather dedup + eliminate need for non-encoded tuple type, and pass start to `proveDataRootTuples`

Reply:
[b9d7da2](https://github.com/celestiaorg/celestia-node/pull/3470/commits/b9d7da2abe995a162e0e9fcd7fc53981f5d261e7)